### ISSUE TITLE
Weirdly specific type narrowing in function call

### ISSUE THREAD

**Issue by tjf801** [microsoft/pyright#4898]
**Describe the bug**
Using generic protocols doesn't type check in certain cases when it should..? I'm honestly not sure what's going on here.

**Expected behavior**
`my_add((1,), (2,))` should type check

**Screenshots or Code**
```python
import time
from typing import Protocol, TypeVar, overload

_T_co = TypeVar("_T_co", covariant=True)
_T_contra = TypeVar("_T_contra", contravariant=True)

class SupportsAdd(Protocol[_T_contra, _T_co]):
    def __add__(self, __value: _T_contra, /) -> _T_co: ...
class SupportsRAdd(Protocol[_T_contra, _T_co]):
    def __radd__(self, __value: _T_contra, /) -> _T_co: ...

@overload
def my_add(a: SupportsAdd[_T_contra, _T_co], b: _T_contra) -> _T_co: ...
@overload
def my_add(a: _T_contra, b: SupportsRAdd[_T_contra, _T_co]) -> _T_co: ...

def my_add(a, b):
    time.sleep(0.1)
    return a + b


my_add((1,), (2,)) # "tuple[Literal[2]]" is incompatible with protocol "SupportsRAdd[tuple[Literal[1]], _T_co@slow_add]"

a: tuple[int] = (1,)
b: tuple[int] = (2,)
my_add(a, b) # same error
```

**Additional context**
I took this example from [this r/python post](https://www.reddit.com/r/Python/comments/10zdidm/why_type_hinting_sucks/)

**Comment by erictraut** [comment]
Pyright is working as designed here. It has to do with pyright's inference rules for tuples, which are documented [here](https://microsoft.github.io/pyright/#/type-inference?id=tuple-expressions). In particular, it infers literal types for the elements in a tuple. This is normally preferable to widening the types to their non-literal forms, but in your example, it's interacting badly with the `SupportsRAdd` generic protocol.
